### PR TITLE
read openstack auth config from client config

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
> // TODO: read/persist client configuration(OS_XXX env vars) in config

/sig openstack

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/assign @dims
**Release note**:

```release-note
try to read openstack auth config from client config and fall back to read from the environment variables if not available
```

### CODE REVIEW

**Comment by dixudx** [comment]
/assign @liggitt 

PTAL. Thanks.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go [hunk 1]
```diff
@@ -145,11 +156,20 @@ func newOpenstackAuthProvider(clusterAddress string, config map[string]string, p
 		}
 	}
 
-	// TODO: read/persist client configuration(OS_XXX env vars) in config
+	getter := tokenGetter{}
+	var authOpt gophercloud.AuthOptions
+	configByte, err := json.Marshal(config)
+	if err == nil {
+		json.Unmarshal(configByte, &authOpt)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: no error handling here?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go [hunk 2]
```diff
@@ -145,11 +156,20 @@ func newOpenstackAuthProvider(clusterAddress string, config map[string]string, p
...
+	}
+	// not empty
+	if !reflect.DeepEqual(authOpt, gophercloud.AuthOptions{}) {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: might be able to just compare straight: `authOpt != gophercloud.AuthOptions{}`

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: needs a test, and an example in doc

Comment (liggitt) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: looks like some options are not serialized in the AuthOptions struct. IdentityEndpoint in particular seems to be ignored. How would the `Token()` method know what server to contact?

Comment (dixudx) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: > How would the Token() method know what server to contact?

L62 `func (t *tokenGetter) Token()` will use this `authOpt` to get an authenticated client, where `IdentityEndpoint` will be used.

Comment (dixudx) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: > looks like some options are not serialized in the AuthOptions struct. IdentityEndpoint in particular seems to be ignored.

Good catch. `AuthOptions` omits this filed. We need to retrieve this value explicitly.

**Comment by dixudx** [comment]
@liggitt Updated. PTAL. Thanks.

**Comment by dixudx** [comment]
> an example in doc (link to PR against 1.10 doc branch)

@dims has already got an in-progress PR kubernetes/website#7304 on this.

**Comment by dims** [comment]
@dixudx please do take that over if you can :)

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go [hunk 3]
```diff
@@ -145,11 +156,33 @@ func newOpenstackAuthProvider(clusterAddress string, config map[string]string, p
 		}
 	}
 
-	// TODO: read/persist client configuration(OS_XXX env vars) in config
+	configByte, err := json.Marshal(config)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: it's not good treat a single config struct as a mix of two API types (we expect a "ttl" key above, and this checks if they happened to specify keys that match the `gophercloud.AuthOptions` struct). what would happen if the `gophercloud.AuthOptions` struct added a ttl field that was an int?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go [hunk 4]
```diff
@@ -145,11 +156,33 @@ func newOpenstackAuthProvider(clusterAddress string, config map[string]string, p
...
+	err = json.Unmarshal(configByte, &authOpt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config to AuthOptions: %s", err)
+	}
+
+	getter := tokenGetter{}
+	// not empty
+	if (authOpt != gophercloud.AuthOptions{}) {
+		// TODO: remove this when gophercloud supports marshal this field
+		endpoint, ok := config["identityEndpoint"]
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: this is another sign that this probably isn't the right approach.

if you want to populate options from inline config, I'd suggest doing it explicitly, like this:

```
gophercloud.AuthOptions{
	IdentityEndpoint: config["identityEndpoint"],
	Username: config["username"],
	Password: config["password"],
	DomainName: config["name"],
	TenantID: config["tenantId"],
	TenantName: config["tenantName"],
}
```

by accepting "identityEndpoint", you are already diverging from a standard auth options json blob, so being explicit doesn't seem bad.

Comment (dixudx) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack.go]: > you are already diverging from a standard auth options json blob, so being explicit doesn't seem bad.

Make sense.

**Comment by dixudx** [comment]
@liggitt Updated. PTAL. Thanks.

**Comment by dims** [comment]
@liggitt ok with the changes based on your comments?

**Comment by dixudx** [comment]
ping @liggitt for another look. Thanks.

**Comment by dims** [comment]
/assign @sttts 

@sttts @liggitt can you please take a look? this seems ready.

**Comment by liggitt** [comment]
/lgtm

needs a better release note and PR with updated documentation on the accepted parameters

**Pull request merged**